libs/@hashintel/component-library/src/lib/flexible-filter.tsx

```diff
@@ -0,0 +1,483 @@
+import { motion, useTransform } from "motion/react";
+import { memo } from "react";
+
+import {
+  calculateDisplacementMap,
+  calculateDisplacementMapRadius,
+} from "./displacement-map";
+import { getDevicePixelRatio } from "./get-device-pixel-ratio";
+import { imageDataToUrl } from "./image-data-to-url";
+import { calculateSpecularImage } from "./specular";
+import { CONVEX } from "./surface-equations";
+import { useToMotion } from "./use-to-motion";
+
+const LATERAL_PART_SIZE = 3; // 3 pixels for top/left/right/bottom parts
+
+// Each part is a Base64-encoded PNG image
+type Parts = {
+  topLeft: string;
+  top: string;
+  topRight: string;
+  left: string;
+  right: string;
+  bottomLeft: string;
+  bottom: string;
+  bottomRight: string;
+};
+
+//
+// Splits an ImageData into 8 parts and returns them as Base64-encoded PNG images.
+// The parts can then be rendered and composited together to form a scalable image with correct corners and edges.
+//
+
+function splitImageDataToParts(
+  imageData: ImageData,
+  radius_: number,
+  pixelRatio: number
+): Parts {
+  const radius = radius_ * pixelRatio;
+  const lateralPartSize = LATERAL_PART_SIZE * pixelRatio;
+
+  if (imageData.width !== radius * 2 + lateralPartSize) {
+    throw new Error("ImageData width is too small for the given radius");
+  }
+  if (imageData.height !== imageData.width) {
+    throw new Error("ImageData should be square");
+  }
+
+  const topLeft = imageDataToUrl(imageData, radius, radius, 0, 0);
+  const top = imageDataToUrl(imageData, lateralPartSize, radius, radius, 0);
+  const topRight = imageDataToUrl(
+    imageData,
+    radius,
+    radius,
+    radius + lateralPartSize,
+    0
+  );
+  const left = imageDataToUrl(imageData, radius, lateralPartSize, 0, radius);
+  const right = imageDataToUrl(
+    imageData,
+    radius,
+    lateralPartSize,
+    radius + lateralPartSize,
+    radius
+  );
+  const bottomLeft = imageDataToUrl(
+    imageData,
+    radius,
+    radius,
+    0,
+    radius + lateralPartSize
+  );
+  const bottom = imageDataToUrl(
+    imageData,
+    lateralPartSize,
+    radius,
+    radius,
+    radius + lateralPartSize
+  );
+  const bottomRight = imageDataToUrl(
+    imageData,
+    radius,
+    radius,
+    radius + lateralPartSize,
+    radius + lateralPartSize
+  );
+
+  return {
+    topLeft,
+    top,
+    topRight,
+    left,
+    right,
+    bottomLeft,
+    bottom,
+    bottomRight,
+  };
+}
+
+//
+// Component that renders the 8 parts of an image and composites them together.
+// Used internally by the Filter component, for DisplacementMap and SpecularMap.
+//
+
+type CompositePartsProps = {
+  imageData: MotionValue<ImageData>;
+  radius: MotionValue<number>;
+  pixelRatio: number;
+  width: MotionValue<number>;
+  height: MotionValue<number>;
+  result: string;
+  hideTop?: boolean;
+  hideBottom?: boolean;
+  hideLeft?: boolean;
+  hideRight?: boolean;
+};
+
+const CompositeParts: React.FC<CompositePartsProps> = memo(
```

Comment:
I wonder - do we need all properties to be motion values? Considering that e.g. `ImageData` is not something that's going to be animated I don't know why it would need to be a `MotionValue`.

Instead of manual memo, I think we should maybe use the react compiler: https://react.dev/reference/react-compiler/compiling-libraries